[JSC] Fuse `instanceof` bytecodes to one to reduce the emitted bytecode size

hyjorc1 opened the pull request:
#### 74d23321eec8326b0df93803a0517a92737f368b
<pre>
[JSC] Fuse `instanceof` bytecodes to one to reduce the emitted bytecode size
<a href="https://bugs.webkit.org/show_bug.cgi?id=280013">https://bugs.webkit.org/show_bug.cgi?id=280013</a>
<a href="https://rdar.apple.com/136316781">rdar://136316781</a>

Reviewed by Yusuke Suzuki.

This patch fuses the emitted bytecodes for JavaScript `instanceof` to one
bytecode. As a result, the emitted bytecode size is reduced from 50 bytes
to 7 bytes for `instanceof`. This change not only improves the space
efficiency of the emitted bytecode size, but also improves execution speed
slightly for `instanceof`.

MicroBenchmark Results with `--inner 20 --outer 20`:

                                     before                    after

instanceof-dynamic               6.3869+-0.0211     ^      6.3034+-0.0135        ^ definitely 1.0132x faster
instanceof-tricky-dynamic        5.5411+-0.0105     ^      5.4403+-0.0133        ^ definitely 1.0185x faster
instanceof-inlining             44.5344+-0.0333     ^     44.4755+-0.0245        ^ definitely 1.0013x faster
instanceof-sometimes-hit         8.2938+-0.0087     ^      8.1762+-0.0072        ^ definitely 1.0144x faster
instanceof-always-hit-two        6.4538+-0.0070            6.4505+-0.0078
instanceof-always-hit-one        2.4569+-0.0079     ^      2.4367+-0.0073        ^ definitely 1.0083x faster
instanceof-bound                19.1643+-0.0201     ?     19.2107+-0.0292        ?

&lt;geometric&gt;                      9.2039+-0.0072     ^      9.1531+-0.0072        ^ definitely 1.0055x faster

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeLivenessAnalysis.cpp:
(JSC::tmpLivenessForCheckpoint):
* Source/JavaScriptCore/bytecode/BytecodeOperandsForCheckpoint.h:
(JSC::valueProfileOffsetFor):
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeLLIntInlineCaches):
(JSC::CodeBlock::tryGetValueProfileForBytecodeIndex):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeFromLLInt):
* Source/JavaScriptCore/bytecode/LLIntPrototypeLoadAdaptiveStructureWatchpoint.cpp:
(JSC::LLIntPrototypeLoadAdaptiveStructureWatchpoint::fireInternal):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitInstanceof):
(JSC::BytecodeGenerator::emitInstanceOf): Deleted.
(JSC::BytecodeGenerator::emitInstanceOfCustom): Deleted.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::InstanceOfNode::emitBytecode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::progressToNextCheckpoint):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::emit_op_instanceof):
(JSC::JIT::emitSlow_op_instanceof):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_instanceof): Deleted.
(JSC::JIT::emitSlow_op_instanceof): Deleted.
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
(JSC::LLInt::handleOpInstanceofCheckpoint):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit_from_inlined_call):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/CommonSlowPaths.h:

Canonical link: <a href="https://commits.webkit.org/284176@main">https://commits.webkit.org/284176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a06693c45c8f34e6be6a9445939c183b4479f8a2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48062 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21329 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72739 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19815 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19631 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/13178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71737 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43919 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59293 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16718 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18172 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/61788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74433 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67918 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12641 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12681 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59373 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/15240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/10211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89697 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43863 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15888 "Found 1 new JSC binary failure: testapi, Found 3867 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_splice_515632.js.default, ChakraCore.yaml/ChakraCore/test/Array/shift_unshift.js.default, ChakraCore.yaml/ChakraCore/test/Basics/ArrayConcat.js.default, ChakraCore.yaml/ChakraCore/test/Boolean/property_and_index_of_boolean.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Error/variousErrors.js.default, ChakraCore.yaml/ChakraCore/test/Function/prototype.js.default, ChakraCore.yaml/ChakraCore/test/Miscellaneous/HasOnlyWritableDataPropertiesCache.js.default, ChakraCore.yaml/ChakraCore/test/Operators/instanceof.js.default, ChakraCore.yaml/ChakraCore/test/RWC/OneNote.ribbon.js.default ... (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/46131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44679 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -224,6 +224,25 @@
     }
 
+# Semantically, this is dst = value instanceof constructor.
+op :instanceof_fused,
```
Given that `instanceof` opcode is removed, let's just rename it to `instanceof`.

Constellation commented on Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
```diff
@@ -352,6 +355,14 @@ void computeUsesForBytecodeIndexImpl(const JSInstruction* instruction, Checkpoin
+        auto bytecode = instruction->as<OpInstanceofFused>();
+        useAt(OpInstanceofFused::getHasInstance, bytecode.m_constructor);
+        useAt(OpInstanceofFused::getPrototype, bytecode.m_value, bytecode.m_constructor, bytecode.m_dst);
+        useAt(OpInstanceofFused::instanceof, bytecode.m_value, bytecode.m_dst);
```
m_dst should not be used for these checkpoints, right?

hyjorc1 replied:
I think we need to use `m_dst` for those checkpoints since we use it for storing the `hasInstance` and `prototype` gotten from `m_constructor`.

Since two `get_by_id`s are needed for getting the `hasInstance` and `prototype`, threes checkpoints are needed
```
op_instanceof_fused:  
<----- getHasInstance
... Get the hasInstance from m_constructor
<------ getPrototype
... Check custom case with the hasInstance
... Get the prototype from m_constructor
<------ instanceof
... Do m_value instanceof the prototype
```

In `DFGByteCodeParser.cpp`, we store the `hasInstance` and `prototype` to the `m_dst`
```
handleGetById(bytecode.m_dst, prediction, get(bytecode.m_constructor), CacheableIdentifier::createFromImmortalIdentifier(hasInstanceImpl), identifierNumber, getByStatus, type, nextCheckpoint());
itermediateIndex = progressToNextCheckpoint();
```

and in `handleOpInstanceofFusedCheckpoint` from `LLIntSlowPaths.cpp` we directly do
```
auto hasInstanceOrPrototype = getOperand(callFrame, bytecode.m_dst);
```

Thus, if we don't use `m_dst` for those checkpoints in `BytecodeUseDef.cpp` then the `hasInstance` and `prototype` won't be propagated back to LLInt once OSR exit happens.

Constellation commented on Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
```diff
@@ -2578,6 +2598,35 @@ static void handleIteratorNextCheckpoint(VM& vm, CallFrame* callFrame, JSGlobalO
+    bool result = JSObject::defaultHasInstance(globalObject, value, hasInstanceOrPrototype);
+    RETURN_IF_EXCEPTION(scope, void());
+    dst = jsBoolean(result);
+    return;
```
This return is not necessary.

Constellation commented on Source/JavaScriptCore/jit/BaselineJITRegisters.h:
```diff
@@ -76,6 +76,13 @@ namespace Instanceof {
+    static constexpr JSValueRegs valueJSR = Instanceof::valueJSR;
+    static constexpr GPRReg constructorGPR = scratch1GPR;
+    static constexpr JSValueRegs hasInstanceJSR = protoJSR;
+}
```
Let's fix this indentation.

Constellation reviewed (approved):
r=me. Let's add fallback C++ implementation in 32bit LLInt.

hyjorc1 committed:
[JSC] Fuse `instanceof` bytecodes to one to reduce the emitted bytecode size
https://bugs.webkit.org/show_bug.cgi?id=280013
rdar://136316781

Reviewed by Yusuke Suzuki.

This patch fuses the emitted bytecodes for JavaScript `instanceof` to one
bytecode. As a result, the emitted bytecode size is reduced from 50 bytes
to 7 bytes for `instanceof`. This change not only improves the space
efficiency of the emitted bytecode size, but also improves execution speed
slightly for `instanceof`.

MicroBenchmark Results with `--inner 20 --outer 20`:

                                     before                    after

instanceof-dynamic               6.3869+-0.0211     ^      6.3034+-0.0135        ^ definitely 1.0132x faster
instanceof-tricky-dynamic        5.5411+-0.0105     ^      5.4403+-0.0133        ^ definitely 1.0185x faster
instanceof-inlining             44.5344+-0.0333     ^     44.4755+-0.0245        ^ definitely 1.0013x faster
instanceof-sometimes-hit         8.2938+-0.0087     ^      8.1762+-0.0072        ^ definitely 1.0144x faster
instanceof-always-hit-two        6.4538+-0.0070            6.4505+-0.0078
instanceof-always-hit-one        2.4569+-0.0079     ^      2.4367+-0.0073        ^ definitely 1.0083x faster
instanceof-bound                19.1643+-0.0201     ?     19.2107+-0.0292        ?

<geometric>                      9.2039+-0.0072     ^      9.1531+-0.0072        ^ definitely 1.0055x faster

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeLivenessAnalysis.cpp:
(JSC::tmpLivenessForCheckpoint):
* Source/JavaScriptCore/bytecode/BytecodeOperandsForCheckpoint.h:
(JSC::valueProfileOffsetFor):
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
(JSC::computeDefsForBytecodeIndexImpl):
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeLLIntInlineCaches):
(JSC::CodeBlock::tryGetValueProfileForBytecodeIndex):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeFromLLInt):
* Source/JavaScriptCore/bytecode/LLIntPrototypeLoadAdaptiveStructureWatchpoint.cpp:
(JSC::LLIntPrototypeLoadAdaptiveStructureWatchpoint::fireInternal):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitInstanceof):
(JSC::BytecodeGenerator::emitInstanceOf): Deleted.
(JSC::BytecodeGenerator::emitInstanceOfCustom): Deleted.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::InstanceOfNode::emitBytecode):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::progressToNextCheckpoint):
(JSC::DFG::ByteCodeParser::parseBlock):
* Source/JavaScriptCore/jit/BaselineJITRegisters.h:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::privateCompileMainPass):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::emit_op_instanceof):
(JSC::JIT::emitSlow_op_instanceof):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_instanceof): Deleted.
(JSC::JIT::emitSlow_op_instanceof): Deleted.
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
(JSC::LLInt::handleOpInstanceofCheckpoint):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit_from_inlined_call):
(JSC::LLInt::llint_slow_path_checkpoint_osr_exit):
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/CommonSlowPaths.h:

Canonical link: https://commits.webkit.org/284176@main
